calblueprint/peoplepower-web
# [1] Set up Redux to Cache User Data

## Setup Redux to Cache User Data

### What Happened in this PR

**I installed redux into the project!**

This has made so much of our code blissfully more simple!

Full info on how it works is here: https://www.notion.so/calblueprint/PP-Power-Redux-0735d9d4783f4100b5d79e180e96c61b

Our store has two main sections: one for user data and one for community
- The user data slice of the store stores: the person record, owner record, userLogin record, projectGroup record, solar project records, credential info, and authentication info
- The community slice of the store stores the announcement array. 

Every main screen has been refactored to get the information it needs from the redux store instead of asking airtable. 

The redux store pulls from airtable 1) when the user logs in 2) when the app loads (if the user was already logged in), 3) when the user updates their profile (currently not working due to a bug) 

**I rearranged some folders**

The lib folder was getting very large so I organized some files within it into sub-folders. I also combined the images and assets folder

### What doesn't work
While I've tested most of the main app's functionality, It's not without bugs. For example, for some reason you can't select a project group in onboarding anymore. Onboarding is sort of a complicated question under redux so I think we should try and figure that out in a separate PR.

### What comes next? 
These things can 100% be tackled during the semester as well as before, just some thoughts as I was refactoring code

- Bring routing into Redux (I plan on adding another PR for this soon)
- Figure out the best way to reduxify onboarding. 
- Refactor Airtable tables to simplify development
- Transform airtable records on read and write at a lower level (in `airtable.js`) instead of in every single component

### How to review this PR

I tried my best to limit the refactoring to what was necessary and making TODO's where I saw problems to fix in the future. But I'm sorry this is a MASSIVE PR. Please ask any and all questions you have and I'm happy to go through it and explain everything over video chat. I genuinely think all of our devs should read through and understand the changes made in this PR since so much of our codebase has been simplified through redux.

Order of which to review files: 
1. Check out the files in `lib/redux/`. The two slice files define the structure of their part of the store as well as the actions available. This is the core of our redux code
2. Check out `userDataUtils.js` for the main data refresh function
3. Check out `Login.js` and `App.js` to see how they "request" a fresh copy of the data when a user logs in 
4. Take a look at a simple component like `Navbar.js` to see how a component can read data from the store using `mapStateToProps`
5. Go through the rest of the components and files to see how they too can be simplified by reading data from redux instead of requesting airtable



### What do I need to install to test it? 

Redux DevTools is a lifesaver while debugging redux: https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en

CC: @aivantg @dfangshuo
